src/schedule/zephyr_ll.c

```diff
@@ -0,0 +1,435 @@
+	 * zephyr_ll_task_free(), see comments there for details.
+	 */
+	list_for_item_safe(list, tmp, &head) {
+		enum task_state state;
+		struct zephyr_ll_pdata *pdata;
+		bool must_run;
+
+		task = container_of(list, struct task, list);
+
+		spin_lock_irq(&sch->lock, flags);
+		if (task->state == SOF_TASK_STATE_PENDING) {
+			task->state = SOF_TASK_STATE_RUNNING;
+			must_run = true;
+		} else {
+			must_run = false;
+		}
+		spin_unlock_irq(&sch->lock, flags);
+
+		/*
+		 * While the lock is not held, the task can be cancelled, we
+		 * deal with it after the task completion.
+		 */
+		if (must_run) {
+			/*
+			 * task's .run() should only return either
+			 * SOF_TASK_STATE_COMPLETED or SOF_TASK_STATE_RESCHEDULE
+			 */
+			state = task_run(task);
+			if (state != SOF_TASK_STATE_COMPLETED &&
+			    state != SOF_TASK_STATE_RESCHEDULE) {
+				tr_err(&ll_tr,
+				       "zephyr_ll_run: invalid return state %u",
+				       state);
+				state = SOF_TASK_STATE_RESCHEDULE;
+			}
+		} else {
+			state = SOF_TASK_STATE_COMPLETED;
+		}
+
+		spin_lock_irq(&sch->lock, flags);
+
+		pdata = task->priv_data;
+
+		if (pdata->terminator) {
+			/*
+			 * zephyr_ll_task_free() is trying to free this task.
+			 * complete it and signal the semaphore to let the
+			 * function proceed
+			 */
+			zephyr_ll_task_done(sch, task);
+			k_sem_give(&pdata->sem);
+		} else if (state == SOF_TASK_STATE_COMPLETED) {
+			zephyr_ll_task_done(sch, task);
+		} else {
+			/*
+			 * task->state could've been changed to
+			 * SOF_TASK_STATE_CANCEL
+			 */
+			switch (task->state) {
+			case SOF_TASK_STATE_CANCEL:
+				zephyr_ll_task_done(sch, task);
+				break;
+			default:
+				/* reschedule */
+				list_item_del(&task->list);
+
+				zephyr_ll_task_insert_unlocked(sch, task);
+
+				task->start = sch->ll_domain->next_tick;
+			}
+		}
+
+		spin_unlock_irq(&sch->lock, flags);
+	}
+
+	notifier_event(sch, NOTIFIER_ID_LL_POST_RUN,
+		       NOTIFIER_TARGET_CORE_LOCAL, NULL, 0);
+}
+
+/* Called once for periodic tasks or multiple times for one-shot tasks */
+static int zephyr_ll_task_schedule(void *data, struct task *task, uint64_t start,
+				   uint64_t period)
+{
+	struct zephyr_ll *sch = data;
+	struct zephyr_ll_pdata *pdata;
+	struct task *task_iter;
+	struct list_item *list;
+	uint64_t delay = period ? period : start;
+	uint32_t flags;
+	int ret;
+
+	tr_info(&ll_tr, "task add %p %pU priority %d flags 0x%x", task, task->uid,
+		task->priority, task->flags);
+
+	irq_local_disable(flags);
+	if (!sch->n_tasks)
+		/*
+		 * If we're adding the first task to the scheduler, it is
+		 * possible that the scheduler has already run before on this
+		 * core and its thread has been aborted while holding the
+		 * spinlock, so we have to re-initialize it. Also note, that
+		 * the counter is decremented as the last operation in
+		 * zephyr_ll_task_done() before aborting the thread, so just
+		 * disabling local IRQs provides sufficient protection here.
+		 */
+		spinlock_init(&sch->lock);
```

Comment:
Why is the spinlock not automatically initialized at core boot as part of scheduler init ?

Reply:
because it has to be initialised every time a core submits the first task for scheduling

Reply:
This is broken and need fixed.

Reply:
the reason is, that when the last task on a core completes, the LL-scheduling thread on that core terminates *while holding the spin-lock*. And the spin-lock is a part of per-core scheduler data, so, next time we start the scheduling thread on that core we re-initialise the spin-lock. So far I don't see a sufficiently clean way to eliminate this... Is it really that bad?

Reply:
Yes, it's bad - lets at least put it int a static inline away from the main flow (for easier reading)

Reply:
you're proposing a
```
static zephyr_ll_init_scheduler_for_first_task(struct zephyr_ll *sch)
{
	spinlock_init(&sch->lock);
}
```
? Yeah, maybe that would self-document it a bit and add a scope to it too.